master/buildbot/test/unit/test_changes_hgpoller.py

```diff
@@ -97,52 +160,18 @@ def test_poll_initial(self):
                        'ssh://example.com/foo/baz')
             .path('/some/dir'),
             gpo.Expect(
-                'hg', 'heads', 'default', '--template={rev}' + os.linesep)
+                'hg', 'heads', '-r', 'default', '--template={rev}' + os.linesep)
             .path('/some/dir').stdout(b"73591"),
-            gpo.Expect('hg', 'log', '-b', 'default', '-r', '73591:73591',  # only fetches that head
-                       '--template={rev}:{node}\\n')
-            .path('/some/dir').stdout(LINESEP_BYTES.join([b'73591:4423cdb'])),
-            gpo.Expect('hg', 'log', '-r', '4423cdb',
-                       '--template={date|hgdate}' + os.linesep + \
-                       '{author}' + os.linesep + \
-                       "{files % '{file}" + os.pathsep + \
-                       "'}" + os.linesep + '{desc|strip}')
-            .path('/some/dir').stdout(LINESEP_BYTES.join([
-                b'1273258100.0 -7200',
-                b'Bob Test <bobtest@example.org>',
-                b'file1 with spaces' + PATHSEP_BYTES +
-                os.path.join(b'dir with spaces', b'file2') + PATHSEP_BYTES,
-                b'This is rev 73591',
-                b''])),
         )
 
         # do the poll
         yield self.poller.poll()
 
         # check the results
-        self.assertEqual(len(self.master.data.updates.changesAdded), 1)
-
-        change = self.master.data.updates.changesAdded[0]
-        self.assertEqual(change['revision'], '4423cdb')
-        self.assertEqual(change['author'],
-                         'Bob Test <bobtest@example.org>')
-        if self.usetimestamps:
-            self.assertEqual(change['when_timestamp'], 1273258100)
-        else:
-            self.assertEqual(change['when_timestamp'], None)
-        self.assertEqual(
-            change['files'], ['file1 with spaces', os.path.join('dir with spaces', 'file2')])
-        self.assertEqual(change['src'], 'hg')
-        self.assertEqual(change['branch'], 'default')
-        self.assertEqual(change['comments'], 'This is rev 73591')
+        self.assertEqual(len(self.master.data.updates.changesAdded), 0)
```

Comment:
Looks like none of the tests process any of the changes. This indicates some kind of failure either in mocking or in the logic of the PR. This should be fixed before the PR can proceed.

Reply:
Well, that's not really true. test_poll_**initial** never should have processed any changes to begin with (as mentioned in [735b981d2807df1b4679a3c009beff0ce7e80731](https://github.com/buildbot/buildbot/pull/4643/commits/735b981d2807df1b4679a3c009beff0ce7e80731)), so that's why it's changed from 1 to 0.

test_poll_**regular** still does process 1 change:
```
        self.assertEqual(len(self.master.data.updates.changesAdded), 1)
        change = self.master.data.updates.changesAdded[0]
        self.assertEqual(change['revision'], '784bd')
        self.assertEqual(change['comments'], 'Comment for rev 5')
```

It is true that there could be better test coverage for the branch and bookmark cases, though. The difficultly in writing gpo.Expect() functions makes this somewhat time-consuming.

Reply:
I've added some additional unit tests.